File: SortVision/public/code/selection/haskell/selectionSort.hs
module SelectionSort (selectionSort) where

-- Time: Best/Avg/Worst O(n^2), Space: O(1) extra (excluding output list)
selectionSort :: Ord a => [a] -> [a]
selectionSort [] = []
selectionSort xs = m : selectionSort (removeFirst m xs)
  where
    m = minimum xs

removeFirst :: Eq a => a -> [a] -> [a]
removeFirst _ [] = []
removeFirst x (y:ys)
  | x == y = ys
  | otherwise = y : removeFirst x ys

main :: IO ()
main = print (selectionSort [64, 25, 12, 22, 11])

File: SortVision/public/code/selection/kotlin/selectionSort.kt
fun selectionSort(arr: IntArray): IntArray {
    // Time: Best/Avg/Worst O(n^2), Space: O(1)
    for (i in 0 until arr.size - 1) {
        var minIdx = i
        for (j in i + 1 until arr.size) {
            if (arr[j] < arr[minIdx]) minIdx = j
        }
        if (minIdx != i) {
            val t = arr[i]
            arr[i] = arr[minIdx]
            arr[minIdx] = t
        }
    }
    return arr
}

fun main() {
    val arr = intArrayOf(64, 25, 12, 22, 11)
    println(selectionSort(arr).joinToString())
}
